### PR TITLE
wip: kms part 1, core encryption functionality and deprecations.

### PR DESCRIPTION
**Pull request by gak**
Related #2290 

Doesn't change any functionality in the system yet. It is just a refactor and new encryption types, with tests.

### CODE REVIEW

**Comment by alecthomas** [comment]
I think it would be preferable not to rename the existing types, it makes it very difficult to tell what has actually changed. Add a [deprecated comment](https://go.dev/wiki/Deprecated) instead. If there are new types call them "*Next" or something for now.

File: internal/encryption/encryption.go [hunk 1]
```diff
@@ -56,88 +62,327 @@ func (n NoOpEncryptor) DecryptJSON(input json.RawMessage, output any) error {
...
-	return nil
+	kekAEAD, err := client.GetAEAD(uri)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get aead: %w", err)
+	}
+
+	return kekAEAD, nil
+}
+
+func NewKmsEncryptorGenerateKey(uri string, v1client *awsv1kms.KMS) (*KmsEncryptor, error) {
```
Comment (alecthomas) [inline internal/encryption/encryption.go:258]: Kms -> KMS in Go

**Comment by alecthomas** [comment]
+ @AlexSzlavik

File: internal/encryption/encryption.go [hunk 2]
```diff
@@ -56,88 +62,327 @@ func (n NoOpEncryptor) DecryptJSON(input json.RawMessage, output any) error {
...
 }
 
-func (e Encryptor) DecryptJSON(input json.RawMessage, output any) error {
-	var payload EncryptedPayload
-	if err := json.Unmarshal(input, &payload); err != nil {
-		return fmt.Errorf("failed to unmarshal encrypted payload: %w", err)
+type SubKey string
+
+const (
+	Logs  SubKey = "logs"
```
Comment (alecthomas) [inline internal/encryption/encryption.go:169]: I think add a `SubKey` suffix to these, just to be clear when browsing the symbols.

**Review by alecthomas — APPROVED** [review]

**Comment by alecthomas** [comment]
> I think it would be preferable not to rename the existing types, it makes it very difficult to tell what has actually changed. Add a [deprecated comment](https://go.dev/wiki/Deprecated) instead. If there are new types call them "*Next" or something for now.

Disregard for this one, but in general this is preferable.

**Comment by gak** [comment]
> > I think it would be preferable not to rename the existing types, it makes it very difficult to tell what has actually changed. Add a [deprecated comment](https://go.dev/wiki/Deprecated)
> 
> Disregard for this one, but in general this is preferable.

Too late!

File: internal/encryption/encryption.go [hunk 3]
```diff
@@ -56,88 +62,327 @@ func (n NoOpEncryptor) DecryptJSON(input json.RawMessage, output any) error {
...
+	bytes, err := primitive.Decrypt(encrypted, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decrypt: %w", err)
+	}
+
+	return bytes, nil
+}
+
+func derivedEncrypt(root keyset.Handle, subKey SubKey, cleartext []byte) ([]byte, error) {
+	// TODO: Deriving might be expensive, consider caching the derived keyset.
```
Comment (stuartwdouglas) [inline internal/encryption/encryption.go:371]: I think we should probably resolve this before it goes in (either we determine that this is cheap enough we don't need a cache, or we should add one).

Comment (gak) [reply internal/encryption/encryption.go:383]: Yeah exactly. I suspect it is too heavy but will measure first.

Comment (gak) [reply internal/encryption/encryption.go:383]: This is part one but I'll check it before the final PR.

**Comment by gak** [comment]
> I think it would be preferable not to rename the existing types, it makes it very difficult to tell what has actually changed. Add a [deprecated comment](https://go.dev/wiki/Deprecated) instead. If there are new types call them "*Next" or something for now.

This is in now. The diff is much nicer.

**Pull request merged**